### PR TITLE
Pin the last Jenkins package which uses sysvinit scripts.

### PR DESCRIPTION
**Pull request by nuclearsandwich**
Earlier this year the packaged versions of Jenkins migrated to systemd units from sysvinit scripts. The current versions of the Jenkins chef cookbook and this cookbook both expect Jenkins to run using sysvinit and make configuration changes using the associated configuration files which are not used by the systemd service.

As a result, newer versions of the Jenkins will not initialize correctly with this cookbook.

Until the upstream Jenkins cookbook is updated to use systemd or we stop using the upstream cookbook entirely, we can't move forward from this Jenkins version.

It is possible for an already-configured farm to continue working after an upgrade and Jenkins cannot be safely downgraded so this version pin is a breaking change and additional work to protect from an unintended downgrade may be warranted.

### CODE REVIEW

**Comment by nuclearsandwich** [comment]
#120 fixes the problem with the repo CI scripts. I may rebase this PR if that merges first.

File: attributes/jenkins.rb
Comment (nuclearsandwich) [inline attributes/jenkins.rb:24]: This is a bit of a carry on from where this change originated (#117) but I'm interested in leaving it in as a signifier for other build farm maintainers. I expect that a PR and release between now and Java 11 support will raise our "high water level" to 2.346.1.

File: recipes/jenkins.rb
Comment (nuclearsandwich) [inline recipes/jenkins.rb:42]: The rest of this gets rough. I'm very sorry but also a little proud.

File: recipes/jenkins.rb
Comment (nuclearsandwich) [inline recipes/jenkins.rb:45]: The `{` opens the Ruby block that chef will eval at convergence time and then within it I immediately start a lambda literal with `-> do` (affectionately known as the "stabby lambda" back in my day) creating a block within a block.

I need to do this in order to use early returns in the code below since using `return` or `break` in a block literal will result in a LocalJumpError. I'll `call` the lambda as soon as I close it passing the result to the outer block.

File: recipes/jenkins.rb
Comment (nuclearsandwich) [inline recipes/jenkins.rb:48]: If Jenkins isn't installed the command above will fail and we can proceed nothing to worry about.

File: recipes/jenkins.rb
Comment (nuclearsandwich) [inline recipes/jenkins.rb]: I'm not actually sure that this code is reachable, I can't think of a situation where the command above would exit successfully but there wouldn't be any version info from it.

I may decide to take this out or turn it into a hard error because it's "wrong"

File: recipes/jenkins.rb
Comment (nuclearsandwich) [inline recipes/jenkins.rb:63]: Earlier attempts at this change tried to just override the node attribute within this block but due to the compile-then-converge nature of chef the attribute value has already been read by the recipe I'm trying to change before this converge-time code runs so instead I put a sigil value in the run_state and get nasty later.

File: recipes/jenkins.rb
Comment (nuclearsandwich) [inline recipes/jenkins.rb:67]: This is calling the lambda literal and closing the `only_if` block value argument. I'm not happy this is necessary but it's the cleanest way I could phrase the code block within and that is what mattered to me.

File: recipes/jenkins.rb
Comment (nuclearsandwich) [inline recipes/jenkins.rb]: By the time we know that we want to, we can't affect the attribute value used by the jenkins cookbook. If we did nothing here, the user would actually get a failure within the Jenkins cookbook when they try to downgrade a package marked for holding. But if we let that happen we risk them not knowing why that is happening and potentially releasing the hold.
So we'll check our sigil value and instead recommend they set the jenkins version that they are currently on which will keep them at their current state until we can make the necessary changes in this cookbook (and the upstream Jenkins one) to unblock further updates.

This is still suboptimal because anyone who is provisioning from scratch using an existing config will experience failures unless they revert back to our new default version for the first provisioning.

File: recipes/jenkins.rb
Comment (cottsay) [reply recipes/jenkins.rb]: Eh, you're parsing output from a command line tool. I think it's totally reasonable to have a protection like this.

Not that you should support the scenario, but I could see something like localization resulting in `nil` here.

File: recipes/jenkins.rb
Comment (cottsay) [inline recipes/jenkins.rb:60]: Could this be cleaned up with the "spaceship operator"?

File: recipes/jenkins.rb
Comment (cottsay) [inline recipes/jenkins.rb]:
```suggestion
        Chef::Log.fatal("Before this cookbook continues, please set the node['jenkins']['master']['version'] attribute to #{node.run_state[:jenkins_package_version_lock]} or this cookbook will attempt to downgrade your Jenkins version.")
```

**Review by cottsay — APPROVED** [review]
Minor comments, but looks good.

Disclaimer: IANAR (I Am Not A Rubist)

File: recipes/jenkins.rb
Comment (nuclearsandwich) [reply recipes/jenkins.rb]: > I could see something like localization resulting in `nil` here.

Yeah, I think you're right and on consideration if we get here we should make the user stop and figure out how and what to do about it. 9fb0a780ad178cffda22568ac67b4d706be9c1bc

File: recipes/jenkins.rb
Comment (nuclearsandwich) [reply recipes/jenkins.rb:60]: `<=>` implements three-way comparison for a class. I don't actually think it helps us unless we parse these version components into an actual class and implement the component level comparisons. Although you could be seeing something tidy that I'm not.

The logic here is complex enough that I was optimizing for plain readability rather than conciseness. I suppose that if we had a Comparable version class here we would be able to write

```
version_installed = Version(version_line.chomp.split(": ")[1].split("."))
if version >= Version([2,332,1])
```

**Pull request merged**
